Allowing flexible folder names and trimming test words from image names

nunogoncalves opened the pull request:
Hi, 

As discussed in #8 I took the liberty to try an implementation that would keep the current behaviour while adding two different possibilities:

- Allow the `folderName` to be set, which will impact the folder in which the images will be saved.
- Allow trimming the `test` prefix of the images names. 

This also impacts the Failure folder.

# Before &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp; After
<img width="224" alt="screenshot 2018-02-25 12 22 12" src="https://user-images.githubusercontent.com/3007012/36641367-86a87626-1a26-11e8-9258-269ae9a750f6.png">&nbsp; &nbsp; &nbsp; &nbsp; <img width="153" alt="screenshot 2018-02-25 12 13 55" src="https://user-images.githubusercontent.com/3007012/36641340-020dd8b6-1a26-11e8-9d41-6542b334560c.png">

In my opinion this makes up for better readability, and also helps referencing custom views. 

You can also create subfolders using `folderName = "folder/subfolder"` which allows you to have:

<img width="193" alt="screenshot 2018-02-25 12 31 45" src="https://user-images.githubusercontent.com/3007012/36641485-dc028f02-1a27-11e8-9e6f-3dcce8d9e1c1.png">

Note:
I have almost no experience with objective-C, so if there's anything that might not be in the best way, accept my apologies and please let me know.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/ios-snapshot-test-case?pullRequest=14) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/ios-snapshot-test-case?pullRequest=14) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/ios-snapshot-test-case?pullRequest=14) it.</sub>

alanzeino commented on FBSnapshotTestCase/FBSnapshotTestCase.h:
```diff
@@ -111,6 +111,17 @@
+ When set, resets the test name to the specified one, which in turn also resets the folder in
+ which the snapshopt is going to be saved.
+ */
+@property (readwrite, nonatomic, assign) NSString *testName;
```
it should be `copy` not `assign`

alanzeino commented on FBSnapshotTestCase/FBSnapshotTestCase.h:
```diff
@@ -111,6 +111,17 @@
+/**
+ If set to true, this will remove the `test` from the image name.
+ */
+@property (readwrite, nonatomic, assign) BOOL trimTestFromImageName;
```
What's the relation of this property to the other one you've added? If you set `testName` to `testThing` and also set `trimTestFromImageName` to `YES` will `testThing` -> `Thing`?

alanzeino commented on FBSnapshotTestCase/FBSnapshotTestCase.m:
```diff
@@ -74,6 +74,27 @@ - (void)setUsesDrawViewHierarchyInRect:(BOOL)usesDrawViewHierarchyInRect
+  _snapshotController.trimTestFromImageName = trimTestFromImageName;
+}
+
+-(BOOL) trimTestFromImageName
```
space

alanzeino commented on FBSnapshotTestCase/FBSnapshotTestCase.h:
```diff
@@ -111,6 +111,17 @@
+/**
+ When set, resets the test name to the specified one, which in turn also resets the folder in
+ which the snapshopt is going to be saved.
+ */
```
One big problem with adding a `testName` property; if you set it after `[super setUp]` it will still work but only until `FBSnapshotVerifyView()` is called. But `FBSnapshotTestController` was written assuming that it would get its `testName` through its initializer. Your headerdoc should mention that `testName` can only be set after `[super setUp]`.

alanzeino replied:
Also add an assert like we use for other properties:

```
  NSAssert1(_snapshotController, @"%s cannot be called before [super setUp]", __FUNCTION__);
```

alanzeino commented on FBSnapshotTestCase/FBSnapshotTestController.m:
```diff
@@ -254,7 +258,9 @@ - (NSString *)_referenceFilePathForSelector:(SEL)selector
                                        identifier:identifier
                                      fileNameType:FBTestSnapshotFileNameTypeReference];
+
   NSString *filePath = [_referenceImagesDirectory stringByAppendingPathComponent:_testName];
```
now that `_testName` is actually a property and not an ivar, rename usage of `_testName` to `self.testName` (except in the initializer)

alanzeino commented on FBSnapshotTestCase/FBSnapshotTestController.m:
```diff
@@ -234,6 +233,11 @@ - (NSString *)_fileNameForSelector:(SEL)selector
   }
 
+  if (_trimTestFromImageName) {
+    NSRange replaceRange = [fileName rangeOfString:@"test"];
```
Will this guarantee to only match the first instance, from the start of the string?

nunogoncalves commented:
@alanzeino @dzenbot thanks for your comments. Will work on them. 👍

alanzeino commented on FBSnapshotTestCase/FBSnapshotTestCase.m:
```diff
@@ -74,6 +74,28 @@ - (void)setUsesDrawViewHierarchyInRect:(BOOL)usesDrawViewHierarchyInRect
+
+-(void)setTrimTestFromImageName:(BOOL)trimTestFromImageName
+{
+  _snapshotController.trimTestFromImageName = trimTestFromImageName;
```
I would guess that this also must be called after `[super setUp]` right?

nunogoncalves replied:
Yeah, right, sorry I missed that one also. Anything that configures the controller must be called only after the controller is instantiated, which is in the setUp method.

nunogoncalves commented on FBSnapshotTestCase/FBSnapshotTestController.h:
```diff
@@ -85,6 +85,16 @@ extern NSString *const FBDiffedImageKey;
 @property (readwrite, nonatomic, copy) NSString *referenceImagesDirectory;
 
+/**
+ The name of the tests. This will be used to create folders with the specified name.
```
I guess there's probably a naming issue around here. 

Internally `testName` refers to the class name rather than the test case itself (if we don't set it, it produces the folder name equal to the name of the test file). So, one thing is for sure, I should definitely change the property name in the `FBSnapshotTestCase` class which is the one available in the API. 

I kept `testName`because it's what was being used internally before, but I guess now, being open to the client it's more meaningful to name it something else. 

What would you think about `folderName`, or even `testClass`... I'm open to suggestions. :) 
I like `folderName` because it's very explicit on what it's going to do. I'm going with that now, and if you'd like something else we can discuss it later. :)

nunogoncalves commented:
Updated the code based on latest comments. 
- Renamed `testName` to `folderName`which I think it's more explicit on its intension
- Updated properties comment information
- Renamed `trimTestFromImageName`to `tripTestPrefixFromImageName` to also be more explicit.

alexissan commented:
Great, thanks!
I noticed that if a identifier contains a slash "/", then a new folder will be created to put the images inside, the problem is that the failure images are not created correctly because prefix are appended to the union of selector and identifier.

Maybe would be great to avoid having slash in the identifier or bad things will happen.
Two things could be done:
- Sanitise identifiers to remove the slash from them
- Throw an error message if identifier contains slash

I had that issue and implemented something similar to this pull request before I found it...

nunogoncalves commented:
Hi @alexissan I just tested your usecase (using the code in this PR) and didn't find any problems. 

Maybe I did something wrong? 

- Set `folderName = "MyCustomView/InnerFolder"`
- Created a snapshot using `recordMode = true`
- This creates the image at "MyCustomView/InnerFolder"
- Set `recordMode = false` and ran tests
- Failing images were created at "FailureDiffs/MyCustomView/InnerFolder" as expected.

alexissan commented:
@nunogoncalves what I mean is pass and identifier like this one:
```swift
let identifier = "myIdentifier/somethingElse"
FBSnapshotVerifyView(view, identifier: identifier)
```
a new folder will be created. can you give a try?

nunogoncalves commented:
@alexissan  what about this:

<img width="386" alt="screenshot 2018-02-25 11 02 19" src="https://user-images.githubusercontent.com/3007012/36640673-733cf75c-1a1b-11e8-9c69-fb3004fa37e5.png">
<img width="235" alt="screenshot 2018-02-25 11 02 28" src="https://user-images.githubusercontent.com/3007012/36640674-735832b0-1a1b-11e8-9b97-363673efd554.png">

If you really want subfolders, I think you can just ignore `identifier` from now on.

alexissan commented:
@nunogoncalves yes, is totally fine. 
I was just saying that using an identifier can introduce problems (unrelated to your fixes).
I like your changes and I hope you can merge very soon.

Thanks so much!!!

nunogoncalves commented:
Hi @alanzeino. Did you get the chance to see the last updates? Do we need any more changes? :)

alanzeino commented:
Okay so I've looked at this again. Part of me is concerned that the PR has evolved beyond the original scope into features that I don't see a use for (but I'm willing to be convinced).

For example, getting the module name out of the folder name; makes a lot of sense. Happy to merge that in. But trimming the `test` prefix, I have less idea what this solves. Same thing with completely custom folder names. I don't know what this helps you with.

In the PR it says these changes help for 'better readability' but I don't see how completely ignoring the class name by overriding it with a different folder name helps — if anything, things get a little more confusing. Same thing with omitting `test` as a prefix. If I'm looking at a bunch of failed snapshots in the `IMAGE_DIFF_DIR`, I would lose the context of what test class generated those snapshots.

Happy to change my mind but for now I would only merge in the change that omits the module name from the folder structure, since I think that is redundant information.

Sorry for not bringing this up earlier.

_P.S., as someone who does massive regenerations of snapshots, not being able to match them with a `test(.+)@2x.png` regex would be annoying. I know this is opt in, but if someone does opt in they'll basically lose the ability to find snapshot PNGs using a simple regex (they'll have to write a more sophisticated script that looks for `ReferenceImages_foo` and considers only PNGs under that dir as snapshot images, for example)._

nunogoncalves commented:
Let me try to sell you my fish. Hopefully not rotten. :) 

Let's say we have a project called `MyAwesomeProject`. Now let's imagine we have a view called `ColorfullView` that we can and want to test. Naturally the name testcase file and class name would be `ColorfullViewSnapshots`. 

With the current implementation, the folder in which the images are going to be saved, will be called `MyAwesomeProject.ColorfullViewSnapshots`. The first thing that comes to mind is the existence of `MyAwesomeProject.` in the folder name. This is exactly what triggered the first pull request #8. 

Now, when I started implementing this update, I didn't simply want to set a flag like `tripModuleName` because that is actually an implementation detail, and as far as I can tell, not a problem in Objective-C. Instead, and this made/makes much more sense in my head, we can actually define the name of the testcase. This gives us the power to choose what fits best for a given test. 

At first the user can simply set the name of the folder to the name of the class and it's done. He can simply define a FBSnapshotTestCase subclass that will be a super class of all test cases, and set  the name there in `setUp()` most likely with `type(of: object)` method. For the most basic usage that's it.

After some getting use to snapshot tests we noticed an interesting thing. It can also work as documentation reference for our views. This is amazing as it is already automatic. If we test a given custom view, we get the benefit of documenting it visually as well. In that context, it makes a lot of sense to call the folder the same name of the view being tested. 

So, now we have the power to move from `MyAwesomeProject.ColorfullViewSnapshots` to `ColorfullView` which improves the experience while glancing at the list of custom views. 

Of course one can name the folder "ClarkKent", but with great power comes great responsability. The user needs to be responsible for what he does. :) I would never expect to find a `redView@2x.png` inside a `ClarkKent` folder.

Subfolders only came into play after some discussion here. I think this fits the `ClarkKent` category, but if it makes sense to someone... 

Considering this documentation scenario, removing the test from the image name also helps a lot. 
Now under `ColorfullView` we could see `red@2x.png` where before we'd read `testRed@2x.png`. 

When you say it prevents people from using regex to find specific cases, you're right, but you also said it's an opt-in feature. It's better to have that option than not having. The lib will allow the user to have that decision rather than preventing him to have the liberty to choose. I could argue that your regex case  is probably far from most people's experiences. But this is not really an argument.

alanzeino commented:
Totally understand the use case you've presented. But I don't think all these changes are worth promoting as features for _this_ repository (more on that later).

There's the issue of stopping people setting `folderName` in individual tests — since `folderName` is something you have described should only be set in `-setUp`, there's nothing stopping someone from having a different name each time. I think that's a little too dangerous to allow, considering the most useful feature we've honed in on is that these features help for documentation, which is not a goal of the project. If I can ever set the folder name, one test class could throw its snapshot files across more than one folder. But if you can solve this, I'll accept the folderName change.

However, I think removing `test` as a prefix is a change we don't want. Again, using the images as documentation is not a goal of the project. In this case, if you wanted to take your snapshots and make them into documentation there's nothing to stop you from removing `test` from all the file names and creating some nice HTML which you host internally at your company that can serve as this documentation. I just don't think this library specifically should focus on constructing documentation itself.

I should apologize that the project doesn't define what its goals are and how we take contributions. At the moment, my goal is to take obviously useful contributions that have minimal risk but large impact. I don't want to merge in changes that can confuse users or which don't meet the goal of making unit testing better by providing a mechanism for UIView comparison testing. I'll improve the documentation to outline this better for future contributors.

In the meantime, I'm totally happy to merge a pull request that removes the swift module name from a folder, and the folderName change _(if you can make it work so that it can only be set once per test class)_. 

For the swift module change, I think it is valid to just match a `.` (period), split on that and return the second part. The only research to be done is if Swift or Objective-C allows a period in a class name (I think they don't).

P.S. about the 'this repository' comment above. I want people to understand that even though we might not accept all changes into the repo, there's a very easy way to keep a fork with changes you want and use it in CocoaPods. Simplest way is to use `:path` in your Podfile and point to a local directory in your repository with the fork. Second most simplest way is to overwrite the `:git` directory and use a specific `:commit` or `:branch` for your changes.

nunogoncalves commented:
Ok @alanzeino thanks for your explanation. We're gonna have to live with that. :) 

I think I found a way to set the folder name and make it only possible on init time. I'll update the PR when I have the time. 

I'll also remove the `trimTestFromImageName´ flag from the PR. 

Also, concerning your last paragraph, that's what we are already doing. Anyway, your HTML idea was food for thought and we'll look into it. One window closed, but a door opened. Thanks again.

nunogoncalves commented:
@alanzeino can you please check the updated code? I have removed the `trimTestFromImageName` and added the possibility to set the folderName, at init time only. 

My objc skills are very low, and apologise for any rookie mistakes. 
Thank you.

alanzeino commented on FBSnapshotTestCase/FBSnapshotTestCase.m:
```diff
@@ -11,17 +11,35 @@
 
 #pragma mark - Overrides
 
+- (instancetype)initWithInvocation:(NSInvocation *)invocation
```
This works in your testing? This initializer is always called when a test case subclass is first initialized?

nunogoncalves replied:
Yes.

alanzeino commented on FBSnapshotTestCase/FBSnapshotTestController.h:
```diff
@@ -85,6 +85,11 @@ extern NSString *const FBDiffedImageKey;
+/**
+ The name of the test. This will be used to create folders with the specified name.
+ */
+@property (readwrite, nonatomic, copy) NSString *testName;
```
this should be read only because the testName is received in the initializer

alanzeino commented on FBSnapshotTestCase/FBSnapshotTestCase.m:
```diff
@@ -11,17 +11,35 @@
+  return self;
+}
+
+- (NSString * _Nonnull)getFolderName
```
`_Nonnull` can be omitted here, it's only needed in the interface or the property definition

nunogoncalves commented:
So just reverted the code on FBSnapshotTestController.m/h
The changes get to be on the interface only.

babbage commented:
Note: the module name is not redundant information in the folder name, if you are using a statically defined folder location to store snapshots that are taken from the same tests being are run against two separate test targets. In these cases, the current module name inclusion automatically ensures unique folder paths, whereas if this was removed you would have collisions in both folder paths and actual filenames.

For some users, this might be a 'feature'. You could now have the same reference images used when tests are run against two test targets.

For other users, this would definitely be a significant regression. I would be in the latter camp.

alanzeino commented:
Yes, what you described is correct. I missed this because at Uber every test bundle has its own reference images directory — to have that collision in our codebase you'd need to have the same test class name used twice in one test bundle.

But I think the norm is to have one giant reference images directory used for many test bundles, so this would probably break in that scenario you described.

maryamaffinityclick commented:
any update on this Pull Request ? 
here I'm having this general class which will check some stuff to record or validate, then if it is recording it will snapshot all screens based on this generic class name. but I wanna define it somewhere in my test case.

nunogoncalves commented:
I don't know why, but I have signed the license agreement twice and it's still pending.

alanzeino commented:
To be honest after @babbage's comment I felt less confident about this change. I'm still not 100% sure this is a good change to make, but I'll think about it a bit more this week and make a decision.

nunogoncalves commented:
I really sold all my arguments. 

There were a few people here who said this is something they'd like to have. It is not a forced feature, it's an opt in. Either way we can always use this fork and that's what I'm doing. 

Thanks for the discussion and keep the good work. :)

nunogoncalves commented:
@alexissan my own. 
The fork with the original implementation of this PR: [working-flexible-names](https://github.com/nunogoncalves/ios-snapshot-test-case/tree/working-flexible-names)

The current pull request lives in another branch: [flexible-names](https://github.com/nunogoncalves/ios-snapshot-test-case/tree/flexible-names)

alanzeino commented:
Okay I'm going to merge this feature. But on second look at the API, the current implementation also doesn't work well. In its current state, this requires overriding a getter in your subclass of `FBSnapshotTestCase`, but that override is called in the initializer of the superclass, which to me results in a side effect of the initialization chain and probably would surprise most users of this library.

The simplest change would be to have the property exposed, and allow it to be set after `[super setUp];` by the subclass (the headerdoc should mention this). 

Also make sure that in the implementation of `-setUp` in `FBSnapshotTestCase.m` that `folderName` is reset to `NSStringFromClass([self class])`. This ensures that an overridden `folderName` doesn't leak into the second test case — I noticed that in testing `initWithInvocation:` is called for each test in a test class but I don't want to rely on this underlying implementation detail.

If anything, there's no need to even set `_folderName` in the initializer — it can be set in `-setUp` in `FBSnapshotTestCase.m`, and then subclasses can override that again by setting the property.

Sorry for all the churn on this change. If you like I can make these changes for you — otherwise I'll wait for your update. Thanks!

nunogoncalves commented:
Thank you Alan. 

I’ll gladdly do it, when I get the chance. I came all the way here, won’t stop now. 🙂

nunogoncalves commented:
Hi @alanzeino 
I was trying to look into this and I think there's a problem. The fact that the the controller is instantiated in the setUp method makes this hard and unexpectedly to accomplish.
Maybe I'm missing something. 

Here's the current implementation of setUp:

```objc
- (void)setUp
{
  [super setUp];
  _snapshotController = [[FBSnapshotTestController alloc] initWithTestName:NSStringFromClass([self class])];
}
```

Considering what you were saying we have the options of setting the folderName before or after instantiating the controller: 

```objc
- (void)setUp
{
  [super setUp];
  self.folderName =  NSStringFromClass([self class])]
  _snapshotController = [[FBSnapshotTestController alloc] initWithTestName: self.folderName;
}
```
Before, obviously fails because we're overriding anything the subclass might have setup.

Next:
 
```objc
- (void)setUp
{
  [super setUp];
  _snapshotController = [[FBSnapshotTestController alloc] initWithTestName:folderName;
  self.folderName =  NSStringFromClass([self class])]
}
```

Since we call [super setUp] and then set the `folderName` property, it won't get the desired effect unless on the subclass sets up `folderName` before calling the `setUp` do:

```swift
override func setUp() {

    self.folderName = String(describing: type(of: component))
    super.setUp()
}
```

It kinda feels strange to me.

What's your take in this? 
Maybe I'm missing something?

alanzeino commented:
Yes you're correct. I guess there's no particularly perfect way to implement this; all the different ways have their individual tradeoffs.

I think the option that works the best would be;

1. `FBSnapshotTestCase` exposes the `folderName` property.
2. `FBSnapshotTestController` no longer takes a `testName` in its initializer and exposes a `folderName` property as well.
3. Whenever the `folderName` on `FBSnapshotTestCase` is set, it updates the `folderName` in `FBSnapshotTestController` as well.

So basically `FBSnapshotTestCase` acts as a intermediary. Making this change will also improve the naming used in `FBSnapshotTestController`; which right now calls the 'folder part' of the full path the `testName` when really the _test name_ is what it calls the `fileName`.

nunogoncalves commented:
@alanzeino could you please re check this? :)

alanzeino commented:
Slight problem with this change at the moment, it changes the file paths for existing tests.

Before:
`ReferenceImages_64/UBAlertControllerSnapshotTests/testActionSheetAlertControllerWithView_2_actions_with_view_11_4@2x.png`

After:
`ReferenceImages_64/testActionSheetAlertControllerWithView_2_actions_with_view_11_4@2x.png`

When you do `stringByAppendingPathComponent:` you need to append both the `folderName` and the `testName`. In the current version, the `testClass` passed in via the default initializer isn't used anywhere.

nunogoncalves commented:
Yeah I was just looking at why the tests failed and noticed it. My tests were performed on my implementation so that worked. I'll fix it. :)

nunogoncalves commented:
Hi @alanzeino 
Have you had the time to recheck this?

alanzeino commented:
You renamed the initializer in FBSnapshotTestController.h, which was a public header in the library.

Breaking changes can be very tricky and difficult to spot in Objective-C.

nunogoncalves commented:
Ah damn! :/ I didn’t think of that as I dont use it. And since the instance was not available through the testcase I didn’t realize that. I’m very sorry.
